src/generic/loading-button/LoadingButton.test.jsx

```diff
@@ -0,0 +1,68 @@
+import React from 'react';
+import { fireEvent, render, waitFor } from '@testing-library/react';
+
+import LoadingButton from '.';
+
+const buttonTitle = 'Button Title';
+
+const RootWrapper = (onClick) => (
+  <LoadingButton label={buttonTitle} onClick={onClick} />
+);
+
+describe('<LoadingButton />', () => {
+  it('renders the title and doesnt handle the spinner initially', () => {
+    const { container, getByText } = render(RootWrapper(() => { }));
+    expect(getByText(buttonTitle)).toBeInTheDocument();
+
+    expect(container.getElementsByClassName('icon-spin').length).toBe(0);
+  });
+
+  it('doesnt render the spinner without onClick function', () => {
+    const { container, getByRole, getByText } = render(RootWrapper());
+    const titleElement = getByText(buttonTitle);
+    expect(titleElement).toBeInTheDocument();
+    expect(container.getElementsByClassName('icon-spin').length).toBe(0);
+    fireEvent.click(getByRole('button'));
+    expect(container.getElementsByClassName('icon-spin').length).toBe(0);
+  });
+
+  it('renders the spinner correctly', async () => {
+    const longFunction = () => new Promise((resolve) => {
+      setTimeout(resolve, 1000);
+    });
+    const { container, getByRole, getByText } = render(RootWrapper(longFunction));
+    const buttonElement = getByRole('button');
+    fireEvent.click(buttonElement);
+    expect(container.getElementsByClassName('icon-spin').length).toBe(1);
+    expect(getByText(buttonTitle)).toBeInTheDocument();
+    // StatefulButton only sets aria-disabled (not disabled) when the state is pending
+    // expect(buttonElement).toBeDisabled();
+    expect(buttonElement).toHaveAttribute('aria-disabled', 'true');
+    await waitFor(() => {
+      // StatefulButton only sets aria-disabled (not disabled) when the state is pending
+      // expect(buttonElement).toBeEnabled();
+      expect(buttonElement).not.toHaveAttribute('aria-disabled', 'true');
+      expect(container.getElementsByClassName('icon-spin').length).toBe(0);
+    });
+  });
```

Comment:
Using setTimeout like this is forcing this function to take 1 second at least, which is about a 100 times slower than it needs to be. 

Instead, you can take control of resolving the promise manually to speed things up. 

First replace `longFunction` with a promise you resolve manually. 

```javascript
let resolve;
    let reject;
    const longFunction = () => new Promise((resolver, rejecter) => {
      resolve = resolver;
      reject = rejecter;
    });
```

Now you have the resolve function outside the scope of the promise.  So you can add the following:

```javascript
await act(async () => resolve());
```

This needs act since it will cause the component to update. Finally you can use the existing final two expect statements, but without the `waitFor`.

Reply:
Thank you @xitij2000! Very clean and elegant.
Solved here: 2c05a817c051da682989f18c0e5282addd2dad61